File: encoding.go
package pact

import "encoding/base64"

// encodeBase64URL encodes bytes to base64url without padding (RFC 4648 §5).
func encodeBase64URL(data []byte) string {
	return base64.RawURLEncoding.EncodeToString(data)
}

// decodeBase64URL decodes base64url (no padding) to bytes.
func decodeBase64URL(s string) ([]byte, error) {
	return base64.RawURLEncoding.DecodeString(s)
}

// ExportBase64URL is the exported version of encodeBase64URL for CLI use.
func ExportBase64URL(data []byte) string {
	return encodeBase64URL(data)
}

// ImportBase64URL is the exported version of decodeBase64URL for CLI use.
func ImportBase64URL(s string) ([]byte, error) {
	return decodeBase64URL(s)
}

File: sdk/python/pact/encoding.py
"""Base64url encoding/decoding per RFC 4648 Section 5, no padding."""

import base64


def encode_base64url(data: bytes) -> str:
    """Encode bytes to base64url string without padding.

    Uses the URL-safe alphabet (A-Z, a-z, 0-9, -, _) with no = padding,
    per RFC 4648 Section 5.
    """
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


def decode_base64url(s: str) -> bytes:
    """Decode a base64url string (with or without padding) to bytes."""
    # Add padding if needed
    padding = 4 - len(s) % 4
    if padding != 4:
        s += "=" * padding
    return base64.urlsafe_b64decode(s.encode("ascii"))
